### DISCUSSION TITLE
Example of event readers?

### DISCUSSION THREAD

**Discussion by drewww** [NSSTC/sim-ecs#76] [Q&A] [answered]
I found the EventBus and I like the idea of tracking entities added and removed systematically like this. But I'm struggling to figure out the right pattern to do this. Specifically I'm not sure where my own callbacks fit in. I can do the following:

```
        const addedReader = actions.eventBus.createReader(EntityAdded);
        actions.eventBus.subscribeReader(addedReader);

         addedReader.execute((e) => {
            console.log("entity added: " + e);
        });
```

But I don't think this is right; that callback never gets called when I subsequently do `.flushCommands()` with a bunch of entities being added.

I've searched around the example code and couldn't find anything salient. Perhaps I'm not meant to interact with the event bus in this way? 

Thanks in advance for any pointers!

**Comment by minecrawler** [comment, marked as answer]
Did you see the [events example](https://github.com/NSSTC/sim-ecs/blob/master/examples/events.ts)? Events usually should be run from systems and use the standard reader and writer.

In order to listen for added entities, you could write a system like this:

```typescript
const EventListenerSystem = createSystem({
    entityAddedEvents: ReadEvents(SimECSAddEntityEvent),
}).withRunFunction(({entityAddedEvents}) => {
    let event;
    for (event of entityAddedEvents.iter()) {
        console.log('Entity Added!', event.entity.id);
    }
}).build();
```

If you really want to create your own reader, you'll want to also listen to the correct event name for added entities (SimECSAddEntityEvent). In your code, you also don't need to use line 2. `eventBus.createReader()` already registers the reader with the event bus.

You could even create your very own event bus outside of the ecs, by instantiating the classes yourself and managing them as you see fit, or implement new classes against the interface in order to switch out parts.

This is unfortunately missing from the docs. I'll try to write up something about it. You can find the internal event definitions [here](https://github.com/NSSTC/sim-ecs/blob/master/src/events/internal-events.ts)

**Reply by drewww** [reply]
Ah! I had seen that example and use my own events internally to pass info between systems. But I missed those internally defined events. That's great -- so much in there that will be useful! 

I definitely don't need/want to create my own EventBus. I just got distracted in those classes thinking that was the way to get at the internal entity/component events I cared about. 

Purely FYI in terms of my sequence of learning here and what docs would be useful. When I started out, I was making entities in a Prepworld, where I could immediately operate on the created entities. I built up a lot of bad patterns working that way. Picking up entity ids and storing them in global state for a player entity, for example. Initializing various display logic immediately. When I started to transition to creating entities in a stage's `activate` method, or using prefabs, all those patterns broke. I was making entities in `actions` which were deferred and even when I flushed a bunch of commands and waited for them to complete, I didn't have obvious immediate handles on what I had just made.

**Reply by minecrawler** [reply]
I actually want to write a "sim-ecs book" (just like there's a Rust and Bevy Book), which guides you through all the topics and contains written explanations akin to the ones in the README. The README is getting way too crowded for all the topics I'd like to explain. It would also explain how to start out, set up the basic mechanisms, contain good coding patterns, etc. Would you be interested in reading the docs that way? It will probably take some time to write, though I'll version it and make it available in this repo.

**Reply by drewww** [reply]
Oh yes! My experience so far has been pretty good, but there are clearly concepts represented in the design of the library that are not immediately obvious to me. Once I "crack" them it feels great and I'm happy with how they work. But they are not always clear up front. For example (and this is obvious in retrospect, but still...) it took me some time to realize that components get passed parameters from `.with` or how `action.commands` worked. I wrote a long question earlier this week about why my entities weren't being created when I called `createEntity` but then in writing it realized it was a deferred action and so of course they weren't immediately available. This stuff is all things you can figure out from the examples and existing docs if you're closing reading, but I can imagine a more systematically laid out version of the primary common patterns for doing things with these tools would be really useful.

**Reply by drewww** [reply]
Oh and the differences in behavior in preptime/runtime worlds -- confusing sometimes when the same methods exist but have different return types and behaviors. I also got stuck for a little while using `getWorld` and getting the preptime variant of my world, updating that, and nothing happening in my runtime world? Or at least I think that's what was going on. Happy to collaborate on an outline of the topics or any writing you do on these topics! I can offer my naive reader eye!

**Reply by minecrawler** [reply]
Thank you for the feedback! If you feel like there's a concept which is hard to grasp, I'd be happy to explain it and add it to the book.

**Reply by drewww** [reply]
Ran into another concept where I could use some more context -- Serialization patterns. I started down the path of writing saving/loading today and pretty quickly hit a wall because the default serializer was failing. I'm not 100% sure why, but it is certainly true that my components are not designed in a way to make this straightforward and they have some complex and custom data structures in them. So I'm not shocked that the defaults fail. I know I need to rework them. But it would be helpful also to have an example of a custom serializer, and some context on how you think about ser/de generally so I approach this in a consistent manner that treads a good line between writing a lot of custom serializer code and restructuring my components so they are simpler. What exactly is required for the default serializers to work? Even my simple `Position` class with two numbers in it is failing.

**Reply by drewww** [reply]
I'm reworking components and it's clear I ... should not be putting logic in components. Prefabs expect simple data types only for "satifies" to work. Which is fine, but I'm curious what the best practices are for the sorts of things I was using methods in components for? 

For example in the Position component I had some bounds checking on a `setPosition` method. Or in some display-related components, I would transform an update like `changeBackgroundColor` into the methods that updated the actual graphical representation for that entity. I can strip all this out, but it feels odd and somewhat dangerous to do this because any system can edit any components values and certain sorts of things should happen every time that happens.

I guess the internal listeners could help this? But I'm not seeing a "component modified" event that I would hook into to process something like "react to the player moving." I had that sort of thing handled in the `setPosition` method. 

I'm probably thinking about this all wrong, but curious if you have a perspective on the "right" ways to do this sort of thing.

**Reply by minecrawler** [reply]
Yes, please do not put any logic at all into components. All logic goes into systems. Components are like database rows in a database.

> bounds checking
> any system can edit any components values and certain sorts of things should happen every time that happens.

In my code base I use a reactive approach for cases where I need logic to happen every time the values change. Which means when the position should be updated, my game logic writes an event (e.g. "MovePositionByEvent"). In order to not delay events by a frame, game logic runs first inside its separate sync-point, then events logic runs to handle e.g. changing position in its own sync-point, then other stuff runs with the updated values (e.g. networking) in another sync-point.

If you really need to have logic on a component, then it's possible to write a custom de-/serializer and supply it when registering the component. There you can instantiate the associated class into an object and fill it with the serialized values. The default implementation will create an object, put all properties from the serialized format on it, and blindly cast it to whatever type it should be (since it expects that the object was correctly serialized before and there's no logic on it). There's some special handling for maps, sets, arrays and dates, though, so they can also become components in theory (I haven't had any practical use for them, yet, though).

> I guess the internal listeners could help this?

Events to react to component modifications are in the backlog. I haven't found any performant way to implement them. They still need a lot of investigation about the API, technical tinkering and work.

> I'm probably thinking about this all wrong

I think you are asking the right questions, and a doc book and cook book could be very helpful.

**Reply by minecrawler** [reply]
btw, sync-points are great since they wait for anything async to finish and then flush commands, however they might be a new concept to you. You can think of the scheduler as a (binary search) tree. When adding a schedule, e.g. with 

```typescript
buildWorld()
  .withDefaultScheduling(root => root.fromPrefab(executionSchedule))
  // ...
```

the `root` parameter actually is the root node of the tree. You can set child nodes on it to run `before` and `after` the root node. Every node is a sync-point, which executes all stages inside and then calls "sync-actions", of which the default one is to flush all commands. You can even add your own actions there, if you really need to do very advanced stuff with the scheduling logic.
In the simplest case, though, you can put everything into the prefab and call it a day:

```typescript
const executionSchedule: ISyncPointPrefab = {
    before: {
        stages: [
            [CounterSystem],
        ],
    },
    stages: [
        [SetterSystem],
    ],
    after: {
        stages: [
            [NetworkSystem],
        ],
    }
};
```

**Reply by drewww** [reply]
That all makes sense, thank you. I will need to noodle on my structure. Sync-points do seem great. 

I'm making a turn-based game, so I'm not performance sensitive. I've experimented a bit with `step()`ing through the systems manually to create this feeling. I didn't quite get the hang of it, but also didn't try very hard since status quo was working fine. So for now I'm still letting the systems run rapidly and just returning instantly if its not their "turn" to run. And then one one of the systems runs every loop to update graphics-relevant logic. However there may be some nice structure to lay in, as you describe, that separates out game logic, model housekeeping, and then display.

**Reply by drewww** [reply]
Is it okay to make use of static methods on components? Will that mess up serialiaztion?

That's a bit of having-my-cake. But it at least lets me encapsulate certain common patterns for updating a component in a logically and physically coherent place.

**Reply by minecrawler** [reply]
Sorry, I just realized I lied :( . I haven't touched the serializers in a while and usually sort my stuff into systems, hence I forgot about it, but taking a look at the [default handlers](https://github.com/NSSTC/sim-ecs/blob/master/src/world/world-builder.util.ts), they are able to instantiate unknown classes. Which is why you need to register components first. So that the class name is known and it can instantiate the correct class and fill its fields. The default handlers should be capable for what you asked.

That means that you can have methods on your components, if you really want to. Just fyi: getters and setters won't be invoked during de-/serialization. [I wrote a test for this](https://github.com/NSSTC/sim-ecs/blob/7cb9f57b23f26f8b8775a0f81e563bdaffb2836c/src/world/world-builder.test.ts#L32-L83) with getter, setter and a public method. The serial representation could look like this: `[{"#RES":1},{"#ID":"1713296169746_27","AComponent":{"_foo":1,"bar":"baz"}}]`

**Reply by drewww** [reply]
Got it, okay. I will look at those examples in more detail. Thank you!

I'm migrating away from having code in my components, that's all fine and sensible. Don't need that; just put a bunch of stuff in there that didn't have any right being a component in the first place. I just didn't know any better. :D 

I may add a custom DeSe class because I'm discovering it's helpful to track elements of my scenegraph linked into the components they draw their display characteristics from. But those should not be written to disk; it's designed that if any of them are missing (as in on initialization or when an entity is added) the render system will repopulate them from scratch just fine. I have a few fields like that in components; they're not game state that needs to be durable.

**Reply by minecrawler** [reply]
You can mark the components which should or should not be written by using an entity query with `save()`. That helps keep the save files lean :)

**Reply by drewww** [reply]
Sorry, I know I'm being chatty in here. Feel free to tell me to slow down if it's too much! But I really appreciate it and your attention here is super helpful! 

I'm experimenting with `save()` and `queryEntities(...)` and it seems like I'm just filtering for entities `With()` and `Without()` certain components. I'm not sure how I would filter at the component level. For example get an entity that has `Player` and `Position` and `Displayable` but don't serialize `Displayable`. 

This is a straightforward pattern to what I don't want to serialize, and I'll poke around at writing my own handler tonight. I suspect that's the simpler answer here. I just don't want to serialize fields with certain types like `Container` or `Texture`. My cursory read of the docs suggest I could easily register fallback handlers for the components that contain problematic types; there are not too many and they share the same logic. I'm not totally clear yet on whether the types those handlers are looking for are the components themselves or the types of fields IN those components. If it's the latter, it's perfect for my needs. If it's the former it's a little more complex but doesn't seem too bad looking at the method signatures.

**Reply by minecrawler** [reply]
It's good, since chatting also helps to review and improve the library :)

Serialization with `save()` currently only serializes entire Entities, there's no way to filter them out. You could write your own de-/serializers. They operate on a component level, but you can, of course, manipulate the field data inside the handlers. The default handlers for data structs I link above may be a good starting point, since they do the bare minimum and are short and easy to understand. Just copy them over, modify them and register them with the components in question.

If you want the filter, though, I might be able to implement it at some point, or you can also create a PR. Maybe the query field could take a ComponentQuery as well and iterate over it (might have to inject a ref to the entity into the query).

**Comment by drewww** [comment]
Trying this now and running into an odd structural issue? I admit I am clueless about modern JS module systems and my googling makes it seem like this is often a build side issue. But I'm not having much luck running it down on that side; everyone else with this issue seem to be doing weirder stuff. Perhaps it will ring a bell for you.

```
import { Actions, ReadEvents, WriteResource, createSystem } from "sim-ecs";
import { SimECSAddEntityEvent, SimECSRemoveEntityEvent } from "sim-ecs/dist/events/internal-events";
```
All my other `sim-ecs` imports look like the first line, and work great. But if I try to import from `internal-events` my IDE suggests the second line. Which causes `vite` to error out:
```
12:30:20 PM [vite] Pre-transform error: Missing "./dist/events/internal-events" specifier in "sim-ecs" package
```

I wonder if this is a consequence of the internal events not being exported from [here](https://github.com/NSSTC/sim-ecs/blob/master/src/index.ts)? Or do you think it's something to do with my own vite-specific build patterns or how I'm importing the classes? 

I last wrote javascript in any depth ~10 years ago so all these tools changed out from under me and I'm rebuilding my knowledge.

**Reply by minecrawler** [reply]
That's true, they aren't exported, but they should be! Quick question: do you use v0.6.4 or are you pulling in the master branch? If you are on master, I can enable you fairly quickly. For the other route, I'll need to check a few things and then publish a new version.

**Reply by minecrawler** [reply]
I added the export on master and also [wrote a test case](https://github.com/NSSTC/sim-ecs/blob/master/src/events/internal-events.test.ts#L8) for it. You can take a look at the code of the test to see the event in action. All tests are green (`npm run test`).

**Reply by drewww** [reply]
I'm using v0.6.4 exact, but I can switch to the master branch to fix this. Heading out for an errand but I'll test it when I'm back!

**Reply by minecrawler** [reply]
If you want a quick fix, switching to master should help. I probably won't be able to publish a new version until next week the earliest, since I have plans for the weekend.

I'm in the CEST timezone (Germany), so it's getting late for me. I'll check back tomorrow.

**Reply by drewww** [reply]
Hmm. I switched to master:

```
    "sim-ecs": "github:NSSTC/sim-ecs#master",
```

And started getting: 

```
sim-ecs.js?v=27d9f96b:1517 Uncaught (in promise) Error: Cannot use a system in two runtimes at the same time!
    at new Xe (sim-ecs.js?v=27d9f96b:1517:17)
    at F.prepareRun (sim-ecs.js?v=27d9f96b:1695:61)
    at application.ts:63:15
```

That line in sim-ecs is [here](https://github.com/NSSTC/sim-ecs/blob/94a925944d664d8e32f6da0dcedc3f1f414f7fb0/src/world/runtime/runtime-world.ts#L139).

Errors on startup, at the line where I call `prepWorld.prepareRun`. When I switch back to v0.6.4 everything works as intended. So I don't think it's a structural issue with calling systems multiple times. I am re-using a system in multiple states, but I think I've seen that work okay in the examples?

To my simple eye it looks like this maybe will always happen? Isn't the context of all systems set twice, once in the prior block and a second time in the block with the error assertion? But if the tests are green I must be understanding what's going on wrong.

**Reply by minecrawler** [reply]
@drewww I switched a change I made earlier to fix another problem on my end. Does the error still occur? If so, can you provide a minimum reproducible example?

**Reply by drewww** [reply]
Yes, LGTM! Thank you!!!

**Reply by drewww** [reply]
Is it intended that using `actions.commands.mutateEntity` causes that entity to appear in the `SimECSAddEntityEvent` and `SimECSRemoveEntityEvent` sets? Not a huge deal but that was an un-expected behavior.

Though I suppose this is a bit of a hacky way to detect ComponentModified-like behavior...

**Reply by minecrawler** [reply]
The mutateEntity command wasn't working as intended and I merged a fix. I also added a new event in order to listen for mutation.
